plugins/opensslcryptoplugin/opensslcryptoplugin.cpp

```diff
@@ -1060,3 +1102,53 @@ Daemon::Plugins::OpenSslCryptoPlugin::aes_decrypt_ciphertext(
     return decryptedData;
 }
 
+QByteArray
+Daemon::Plugins::OpenSslCryptoPlugin::aes_auth_encrypt_plaintext(
+        Sailfish::Crypto::CryptoManager::BlockMode blockMode,
+        const QByteArray &plaintext,
+        const QByteArray &key,
+        const QByteArray &init_vector,
+        const QByteArray &auth,
+        QByteArray *tagData)
+{
+    if (auth.isEmpty() || !tagData) {
+        qWarning() << "Authenticated encrypt failed: auth or tag not set!";
+        return QByteArray();
+    }
+
+    if (blockMode == Sailfish::Crypto::CryptoManager::BlockModeGcm
+            && init_vector.size() != SAILFISH_CRYPTO_GCM_IV_SIZE) {
+        qWarning() << "Authenticated encrypt failed: IV size is" << init_vector.size()
+                   << "but should be" << SAILFISH_CRYPTO_GCM_IV_SIZE;
+        return QByteArray();
+    }
+
+    QByteArray encryptedData;
+    unsigned char *encrypted = NULL;
+    unsigned char *tag = NULL;
+    int encryptedSize = -1;
+    const int tagSize = SAILFISH_CRYPTO_GCM_TAG_SIZE;
+
+    encryptedSize = osslevp_aes_auth_encrypt_plaintext(getEvpCipher(blockMode, key.size()),
+                                                       (const unsigned char *)init_vector.constData(),
+                                                       (const unsigned char *)key.constData(),
+                                                       key.size(),
+                                                       (const unsigned char *)auth.constData(),
+                                                       auth.size(),
+                                                       (const unsigned char *)plaintext.constData(),
+                                                       plaintext.size(),
+                                                       &encrypted,
+                                                       &tag,
+                                                       tagSize);
+
+    if (encryptedSize <= 0) {
```

Comment:
Doesn't `-1` from `osslevp_aes_auth_encrypt_plaintext` indicate an error?

Reply:
Yes, so this returns a null QByteArray in that case.

Reply:
Shouldn't there be an indication of success/failure to the caller? Or does an empty `QByteArray` always mean failure?

Reply:
Ideally, all main points of failure would have been checked before calling this aes_auth_encrypt_plaintext() function, so that the only real error that occurs here are to do with a failure to encrypt from the openssl libraries, which is indicated by the empty QByteArray. But, I think having the various checks in this method before calling osslevp_aes_auth_encrypt_plaintext() make this less clear, so I'll change it to move the checks to the calling execEncrypt() method.

Later on, we should probably change the evp.c methods to return appropriate error codes instead of just a -1 on error, to provide more precise error information.

Reply:
Patch updated to clean up the argument checking before calling aes_auth_encrypt_ciphertext() and aes_auth_decrypt_ciphertext() and return more precise errors.

Reply:
Cool, thank you. Agreed about updating evp.c later.